src/protocol/secureMessages.ts

```diff
@@ -0,0 +1,356 @@
+ *    an encrypted request. This will return an ephemeral public key,
+ *    which is used to encrypt the next request.
+ * 2. Encrypted requests are *encrypted* (obviously) and from a Lattice
+ *    protocol perspective they are all constructed the same way:
+ *    create a buffer of `payload` length and fill it with unencrypted
+ *    data, then encrypt the entire payload (not just the data you filled)
+ *    with the ECDH secret formed from the last ephemeral public key.
+ *    The response to this request will contain a new ephemral public
+ *    key, which you will need for the next encrypted request.
+ */
+import {
+  Client
+} from '../index'
+import {
+  ProtocolConstants as Constants,
+  LatticeSecureMsgType,
+  LatticeMsgType,
+  LatticeProtocolVersion,
+  LatticeSecureEncryptedRequestType,
+} from './latticeConstants';
+import {
+  aes256_decrypt,
+  aes256_encrypt,
+  checksum,
+  getP256KeyPairFromPub,
+  randomBytes,
+} from '../util';
+import {
+  getEphemeralId,
+  request,
+} from '../shared/functions';
+import {
+  validateEphemeralPub,
+} from '../shared/validators';
+
+const { msgSizes } = Constants;
+const { secure: szs } = msgSizes;
+
+/**
+ * Build and make a request to connect to a specific Lattice
+ * based on its `deviceId`.
+ * @param client - Instance of (unconnected) client
+ * @param deviceId - Device ID for the target Lattice. Must be in
+ *                   the same `client.baseUrl` domain to be found.
+ * @return {Buffer} - Connection response payload data, which contains
+ *                    information about the connected Lattice.
+ */
+export async function connectSecureRequest(
+  client: Client,
+  deviceId: string,
+): Promise<Buffer> {
+  // Build the secure request message
+  const payloadData = serializeSecureRequestConnectPayloadData(client);
+  const msgId = randomBytes(4);
+  const msg = serializeSecureRequestMsg(
+    client, 
+    msgId, 
+    LatticeSecureMsgType.connect, 
+    payloadData
+  );
+  // Send request to the Lattice
+  const url = `${client.baseUrl}/${deviceId}`;
+  const resp = await request({ url, payload: msg });
+  if (resp.length !== szs.payload.response.connect - 1) {
+    throw new Error('Wrong Lattice response message size.');
+  }
+  // Pack into struct and return data
+  const respStruct: LatticeSecureConnectResponsePayload = {
+    data: resp,
+  }
+  return respStruct.data;
+}
+
+/**
+ * Build an encrypted secure request using raw data,
+ * then send that request to the target Lattice, handle
+ * the response, and return the *decrypted* response
+ * payload data.
+ * Also updates ephemeral public key in the client.
+ * This is a wrapper around several local util functions.
+ * @param client - Instance of Client
+ * @param data - Unencrypted raw calldata for function
+ * @param requestType - Type of encrypted reques to make
+ * @return {Buffer} Decrypted response data (excluding metadata)
+ */
+export async function encryptedSecureRequest(
```

Comment:
Should this be called "request" when it is also serializing and decrypting?

Reply:
Open to suggestions